plugins/occt/module/vtkF3DOCCTReader.cxx

```diff
@@ -285,6 +315,104 @@ class vtkF3DOCCTReader::vtkInternals
+  {
+    StyleMap inheritedStyles;
+
+    if (rootLabel.IsNull())
+    {
+      return inheritedStyles;
+    }
+
+    /* collect styled shapes from the document */
+    StyleMap collectedStyles;
+    XCAFPrs::CollectStyleSettings(rootLabel, TopLoc_Location(), collectedStyles);
+
+    /* iterate styled shapes and assign style if leaf, otherwise collect if parent node */
+    const TopAbs_ShapeEnum leafType = this->Parent->GetReadWire() ? TopAbs_EDGE : TopAbs_FACE;
+    std::vector<std::pair<TopoDS_Shape, XCAFPrs_Style> > parents;
+    for (StyleMap::Iterator iter(collectedStyles); iter.More(); iter.Next())
+    {
+      const TopoDS_Shape& shape = iter.Key();
+      if (shape.ShapeType() == leafType)
+      {
+        inheritedStyles.Add(shape, iter.Value());
+      }
+      else if (shape.ShapeType() < leafType)
+      {
+        parents.emplace_back(std::make_pair(shape, iter.Value()));
+      }
+    }
+
+    /* sort parents from deepest upwards */
+    std::sort(parents.begin(), parents.end(),
+      [](const std::pair<TopoDS_Shape, XCAFPrs_Style>& a,
+        const std::pair<TopoDS_Shape, XCAFPrs_Style>& b)
+      { return a.first.ShapeType() > b.first.ShapeType(); });
+
+    /* pass down each parent style props to descendent edge/face leaves */
+    const auto passDownToLeaves = [&](TopAbs_ShapeEnum type)
+    {
+      for (const auto& parent : parents)
+      {
+        for (TopExp_Explorer iter(parent.first, type); iter.More(); iter.Next())
+        {
+          try
+          {
+            this->PassDownStyleProps(parent.second, inheritedStyles.ChangeFromKey(iter.Current()));
+          }
+          catch (Standard_NoSuchObject&)
+          {
+            inheritedStyles.Add(iter.Current(), parent.second);
+          }
+        }
+      }
+    };
+
+    passDownToLeaves(TopAbs_FACE);
+
+    if (this->Parent->GetReadWire())
+    {
+      passDownToLeaves(TopAbs_EDGE);
+    }
+
+    /* pass down default style (if any) to all leaves */
+    try
+    {
+      const XCAFPrs_Style& defaultStyle = collectedStyles.FindFromKey(rootShape);
+      for (StyleMap::Iterator iter(inheritedStyles); iter.More(); iter.Next())
+      {
+        XCAFPrs_Style style = iter.Value();
+        this->PassDownStyleProps(defaultStyle, style);
+      }
+    }
+    catch (Standard_NoSuchObject&)
+    {
+      /* ignore */
+    }
+
+    return inheritedStyles;
+  }
+
+  //----------------------------------------------------------------------------
+  void PassDownStyleProps(const XCAFPrs_Style& parent, XCAFPrs_Style& child)
+  {
+    if (!child.IsSetColorCurv() && parent.IsSetColorCurv())
+    {
+      child.SetColorCurv(parent.GetColorCurv());
+    }
+
+    if (!child.IsSetColorSurf() && parent.IsSetColorSurf())
+    {
+      child.SetColorSurf(parent.GetColorSurfRGBA());
+    }
+
+    if (child.Material().IsNull() && !parent.Material().IsNull())
+    {
+      child.SetMaterial(parent.Material());
```

Comment:
do we need new data to reach this codepath ?

Reply:
that would be nice yeah. Guess I'll need to learn some more OCCT then

Reply:
We may not be able to hit this with STEP/IGES files as apparently they only support basic colors (or at least the OCC step exporter only exports colors)

https://www.opencascade.com/blog/pbr-in-occt-3d-viewer/
> - XBF document stores any combination of material and color labels. It should be noted, however, that assigning both color and material to the same part gives not well defined result and should be avoided.
> - STEP, IGES and similar formats storing only color information keep using a color table in an XCAF document. Visualization materials in the document are exported as colors.
> - JT reader/writer now translates material properties into a Common material definition in an XCAF document.

no idea what XBF and JT are and whether we can use them or not

Reply:
fine if we cant, as you see fit.